storage/src/lib.rs

```diff
@@ -71,10 +71,13 @@ pub struct BlockHeaderWithHash {
 impl BlockHeaderWithHash {
     /// Create block header extensions from plain block header
     pub fn new(block_header: BlockHeader) -> Result<Self, MessageHashError> {
-        Ok(BlockHeaderWithHash {
-            hash: block_header.message_hash()?.try_into()?,
-            header: Arc::new(block_header),
-        })
+        let hash = if let Some(hash) = block_header.hash().as_ref() {
+            hash.as_slice().try_into()?
+        } else {
+            block_header.message_hash()?.try_into()?
+        };
+        let header = Arc::new(block_header);
```

Comment:
@akoptelov @tizoc 
I am still thinking about this structure,
for example everywhere, we construct BlockHeaderWithHash, we need to clone BlockHeader data:
```
BlockHeaderWithHash::new(message.current_block_header().clone())?;
```
and I think it is also wastefull
for example, what if all messages like: `message.current_block_header`, has `Arc<BlockHeader>` instead of `BlockHeader`, what do you think?

Reply:
Then we need to redesign messages completely. Not something that can be done as a quick fix.